### ISSUE TITLE
Indent fails

### ISSUE THREAD

**Issue by malthe** [abingham/ponylang-mode#9]
I'm on GNU Emacs 24.5.1 – even with ``emacs -q`` – and this mode does not indent at all. I tried putting a breakpoint with ``M-x debug-on-entry ponylang-indent-line`` but that didn't seem to trigger anything. I might be doing it wrong though.

**Comment by abingham** [comment]
Is it doing syntax highlighting?

**Comment by malthe** [comment]
yes!

**Comment by abingham** [comment]
Hmmm. I've tried the most recent melpa package on 24.5.1, and it seems to work. (Note that I'm going to update it soon, because "work" in this case is pretty limited!) 

A few questions, and I apologize if they're really basic/obvious:
1. Are you sure the buffer is in `ponylang-mode`? Try `M-x describe-mode` to check.
2. Are you sure you have the most recent `ponylang-mode` package?
3. In a `ponylang-mode` buffer, what is the value of the variable `indent-line-function`? Try `M-x describe-variable indent-line-function`. It should be `ponylang-indent-line`.

If everything checks out there, then I'm really puzzled.

**Comment by malthe** [comment]
Yep, all three check out. I don't understand ``M-x debug-on-entry`` isn't working for me, but ``ponylang-indent-line`` is indeed being called on ``<TAB>`` – I put a ``(debug)`` to explicitly verify this.

I think the following is a little buggy:
```lisp
(if (looking-at "^[ \t]*\\(repeat\\|until\\|while\\|let\\|for\\|be\\|new\\|use\\|var\\|try\\|else\\|end\\|if\\|ref\\|then\\|fun\\|tag\\)")
  (progn
    (setq cur-indent (+ (current-indentation) tab-width))
    (setq not-indented nil))
(if (bobp)
    (setq not-indented nil)))))))
```
Like if any of those keywords match then it indents the line? That seems wrong.

I did some research and looked at ``elixir-mode``. It's using syntax tables and `smie` – I think it might be a better foundation? But I'm an elisp newbie.

It does seem like the Elixir mode is a good place to look for inspiration. It has tests which basically amount to dropping some code with zero indentation, asking the mode to indent it and verify the result. Their tests have some DRY violations which can easily be avoided:
```lisp
(elixir-def-indentation-test indent-function-calls-without-parens
                             (:tags '(indentation))
  "
test \"foo\" do
assert true, \"should be true\"
assert !false, \"should still be true\"
end
"
  "
test \"foo\" do
  assert true, \"should be true\"
  assert !false, \"should still be true\"
end
")
```
I would probably go remove indentation from the expected result and then apply the indentation routine.

**Comment by abingham** [comment]
You're right that there are almost certainly defects in how the indentation is handled right now. The fact is that I don't really use pony much, and also I'm pretty new at writing modes. In any event, I've pushed a new change that should improve things a bit.

**Comment by malthe** [comment]
Okay let me try out the change. Thanks!

I am somewhat curious to try and write an editing mode myself so I might have a go at it to see if I can adapt the Elixir mode patterns without digging myself into a week-long hole.

**Comment by abingham** [comment]
I'd be happy to get any help I can, esp. if you're using pony on a regular
basis. If there are better, more idiomatic ways to do what I'm doing,
that's great.

**Comment by colin-adams** [comment]
I just tried indent-region, and it just seems to indent each line by one more than the previous, irrespective of the syntactical context.
Which means I have to indent every line by hand.

**Comment by abingham** [comment]
I've just pushed a fix that will alleviate some of the issues, but there are certainly still some problems. Honestly, indenting pony correctly seems fraught with issues, and I think that automatic indentation will never be 100% perfect. For example, what indentation do I use after a line that says "end"? As far as I can tell there's no right answer.

With that said, e.g. Python mode does a pretty good job, and it has similar problems. We just need smarter heuristics, I think.

What would really help me is specific cases where there are problems.

**Comment by colin-adams** [comment]
Now indent-region doesn't work at all - I get:

Wrong type argument: number-or-marker-p, nil

for all files that I try it on.

Using tab key works as before (repeated applications tab back or forward in
some kind of cycle)

**Comment by abingham** [comment]
Oops, sorry about that. I merged in someone's patchset without thoroughly
testing it, and it had a small bug. Try it now.

**Comment by malthe** [comment]
@abingham with pony unlike python you don't have the ambiguity as to when a block ends.

**Comment by colin-adams** [comment]
Working again, much as before.

Here is an example from the tutorial, after calling indent-region on the whole file:

// Define a suitable type
class Pair
  var _x: U32 = 0
  var _y: U32 = 0
  
  new create(x: U32, y: U32) =>
    _x = x
    _y = y



    // Define a + function
    fun add(other: Pair): Pair =>
      Pair(_x + other._x, _y + other._y)

      // Now let's use it
      class Foo
        fun foo() =>
          var x = Pair(1, 2)
          var y = Pair(3, 4)
          var z = x + y

Problems that I see, and heuristics I think that would tackle them:

1) fun is indented one tab too many - Always indent fun with one tab from the margin
2) class is indented by two tabs - Always align class with the left margin

**Comment by abingham** [comment]
@malthe You're correct, of course, but the determination of how to indent after an "end" is not simple. In some cases you want to use the same identation as the "end":
```
while j < 8 do
. . .            
end
var . . .
```
and in other cases you want one level fewer:
```
  var view: Array[U8] iso =
    recover
      Array[U8]((y - x) * (width >> 3))
    end

  let group_r = Array[F32].undefined(8)
```

I misspoke when I said that pony has issues similar to python, but we'll need some relatively sophisticated logic to know how to indent. At this point we just don't have it.

**Comment by colin-adams** [comment]
@abingham I doubt if indentation of var and let should depend upon a preceding end. And having while indented to the left margin looks wrong to me.

**Comment by abingham** [comment]
@colin-adams It doesn't depend solely upon a preceding `end`, but currently if I see an `end` I try to use it. This may be a serious mistake; perhaps I should ignore `end`s and rely on other syntax markers.

And yes, that `while` is out of context for demonstration purposes. IRL it would be in some realistic indentation context. Sorry for the confusion.

**Comment by abingham** [comment]
@colin-adams Regarding your heuristics, the one or classes sounds spot on. As far as I can tell a class can only be defined at that identation level.

I'm not so sure about `fun`, however. It [seems to be possible to have functions at higher indentation levels](https://github.com/CausalityLtd/ponyc/blob/master/examples/readline/main.pony#L44).

**Comment by colin-adams** [comment]
Are, right.
So fun needs to be indented one more than the enclosing type-former keyword
(class, object, actor, anything else?)

**Comment by colin-adams** [comment]
primitive also (?)

**Comment by abingham** [comment]
@colin-adams Right, that sounds like the outline of a good strategy for `fun`. (What has my life come to that I need strategies for fun!?) I'm not sure when I'll have time to really bear down on this, so please feel free to take a try if you're so inclined. I think some of what's in the mode already should serve as a reasonable guide, and I'm happy to help out.

**Comment by colin-adams** [comment]
@abingham I'll just have to be patient then (I've tried editing emacs mode for Eiffel in the past, and i ended up baffled).
